[DebugInfo] Adding DIExpression into SIL

mshockwave opened the pull request:
This PR adds DIExpression infrastructure into SIL.

Previously, we can not use `debug_value` to mark the updated (SIL) SSA of a sub-field inside an aggregate-type source variable. For example:
```swift
struct MyStruct {
    var x : Int
}
%2 = alloc_stack $MyStruct, name "my_struct", loc "file.swift":3:4, scope 3
...
%5 = struct_element_addr %2 : $*MyStruct, #MyStruct.x, loc ...
```
In this snippet we are not able to generate a correct LLVM debug info for `%5` using `debug_value` or `debug_value_addr`. Because the current SIL debug infrastructure lacks the equivalent feature of LLVM's `!DIExpression`. More specifically, the `DW_OP_fragment` operator.

This is not a problem in `-Onone`, but an outstanding one in optimized build where optimizations -- SROA for example -- might break an aggregate-type object apart or even eliminate the original aggregate type. In which case the debug info quality for variable values will degrade.

This PR (when it's finished) essentially adds functionalities similar to `!DIExpression` at the SIL level. Take the code above as the example, we're now able to generate the following SIL:
```swift
struct MyStruct {
    var x : Int
}
%2 = alloc_stack $MyStruct, name "my_struct", loc "file.swift":3:4, scope 3
...
%5 = struct_element_addr %2 : $*MyStruct, #MyStruct.x, loc ...
debug_value_addr %5 : $*Int, var, (name "my_struct", loc "file.swift":3:4, scope 3), type $MyStruct, expr op_fragment:#MyStruct.x, loc "file.swift":10:11, scope 3
```
Here are the new parameters / features added to `debug_value` / `debug_value_addr`:
 - The debug variable can have (optional) SILLocation and/or SILDebugScope attached (the `loc` and `scope` directives _inside_ the parens enclosing `name`), which represents the source location where it was originally declared. If this extra info is absent, the instruction's source location will be used.
 - The `type` directive can be used to indicate the type where the debug variable was originally declared. If this info is absent, the type of the SSA value (i.e. `%5` in this case) will be used.
 - Finally, the `expr` directive is used to carry the DIExpression that basically has a one-to-one mapping to LLVM's `!DIExpression`. Currently we only have `op_fragment` and `op_deref` operator, corresponding to `DW_OP_LLVM_fragment` and `DW_OP_deref`, respectively.

The part that is slightly controversial is probably how we carry these extra information on in-memory representation for the affected SIL instructions. This change will make `SILDebugVariable`, the class that represents debug variable and used by all debug-info-carrying instructions (including the `alloc_*` family), more expensive. Luckily most of the time `SILDebugVariable` is generated on-demand (it is actually not _stored_ inside aforementioned SIL instructions) so hopefully it won't affect conventional SIL instruction transformations (like copying or moving SIL instructions).

I'm splitting this PR into two commits: The first commit add DIExpression support into SIL instructions classes (e.g. `DebugValueInst`) and changes regarding SILParser / SILPrinter; The second one add IRGen support.

mshockwave committed:
[SIL][DebugInfo][Patch 1/2] Initial SIL DIExpression support

Currently the debug info infrastructure inside SIL can only associate a
source variable with a single (simple) SSA value. Which is insufficient
to preserve (correct) debug info across SIL-level optimizations -- for
example, SROA that decompose or even eliminate aggregate-type obejcts.

By adding DIExpression into SIL, we are able to reconstruct the
connection between a source variable and its SSA value counterpart, even
across optimizations. This patch adds such support into in-memory
representation for SIL instructions and the SILParser/Printer. The
following patch will add changes for the IRGen part.

mshockwave commented:
I'll add test cases shortly

adrian-prantl commented on include/swift/SIL/SILDebugInfoExpression.h:
```diff
@@ -0,0 +1,121 @@
+  Kind getKind() const { return OpKind; }
+
+  SILDIExprOperator getAsOperator() const {
+    return OpKind == K_Operator? Operator : SILDIExprOperator::INVALID;
```
clang-format

adrian-prantl commented on include/swift/SIL/SILDebugInfoExpression.h:
```diff
@@ -0,0 +1,121 @@
+struct SILDIExprOperand {
+  enum Kind {
+    K_Operator,
+    K_Decl
```
Are we using a K_ prefix elsewhere? I would probably just drop it.

mshockwave replied:
Since there is also a member variable called "Operator" in this class, I'll rename `K_Operator` into `OperatorKind` to avoid ambiguous (so does `K_Decl`).

adrian-prantl commented on include/swift/SIL/SILDebugInfoExpression.h:
```diff
@@ -0,0 +1,121 @@
+namespace swift {
+class TailAllocatedDebugVariable;
+
+enum class SILDIExprOperator : unsigned {
```
It would be great to add Doxygen comments to all new classes.

adrian-prantl commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -4540,9 +4566,18 @@ class MarkFunctionEscapeInst final
+  unsigned NumDIExprOperands;
+  bool HasAuxDebugVariableType;
+  // bit 0: Has debug location
+  // bit 1: Has debug scope
```
`///`

adrian-prantl replied:
Maybe define an enum instead of commenting?

adrian-prantl commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -4583,9 +4656,18 @@ class DebugValueInst final
+  bool HasAuxDebugVariableType;
+  // bit 0: Has debug location
+  // bit 1: Has debug scope
+  unsigned HasAuxVariableSourceLoc : 2;
```
and reuse the same enum here?

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2330,9 +2337,61 @@ void IRGenDebugInfoImpl::emitArtificialFunction(IRBuilder &Builder,
+  const SILDIExprOperand &CurOp = *It++;
+  assert(CurOp.getAsOperator() == SILDIExprOperator::Fragment);
+  // Expecting a VarDecl that points to a field in an struct
+  auto *VD = dyn_cast_or_null<VarDecl>((It++)->getAsDecl());
```
In DIExpression, we define a custom iterator to decode the uint64_t array. Maybe we want to copy this?

mshockwave replied:
I'm not sure I understand you correctly: SILDebugInfoExpression doesn't store a list of uin64_t but some high-level objects like `Decl*`, so I don't think we can reuse llvm::DIExpression's iterator here.

adrian-prantl replied:
Would it make sense to also define a (new) iterator _similar_ to DIExpression's that makes it convenient to iterate through the list of SILDIExprOperand one operator at a time. So operator++ would skip over the entire Fragment, including its two parameters.

mshockwave replied:
That is actually a good idea! Will add this feature

adrian-prantl commented on lib/SIL/IR/SILDebugInfoExpression.cpp:
```diff
@@ -0,0 +1,27 @@
+//===--- SILDebugInfoExpression.cpp - DIExpression for SIL ------*- C++ -*-===//
```
The `-*- C++ -*-` marker only makes sense in a `.h` file, where it's ambiguous whether the language is C or C++ based on the file name.

mshockwave commented:
I'll squash all the commits into one right before it's ready to merge (or I can do it right now if you prefer)

mshockwave commented:
@swift-ci  test

mshockwave commented:
@swift-ci please test

mshockwave committed:
[IRGen][Patch 2/2] Add IRGen support for SIL DIExpression

This patch is primarily doing two things:
 - Teach IRGen for some of the instructions to generate debug info based
   on SILDebugVariable rather than AST node, which is not always
   available when reading from SIL.
 - Generate corresponding `llvm.dbg.*` intrinsics and `!DIExpression`
   from the newly added SIL DIExpression.

mshockwave commented:
@swift-ci please test

mshockwave commented:
Add documentations
@swift-ci please test

adrian-prantl commented:
Thanks, this is looking great! It would be good to also have test cases for the Verifier assertions that are added by this patch.

adrian-prantl commented:
@jckarter Do you have any opinion on the proposed SIL syntax extension?

mshockwave commented:
> Thanks, this is looking great! It would be good to also have test cases for the Verifier assertions that are added by this patch.

Sorry I almost missed this comment.
And hopefully the windows build bot is back to work 😛 

@swift-ci please test

adrian-prantl reviewed (approved):
LGTM.
